### PR TITLE
#473 Refactoring Request-Line validation

### PR DESCRIPTION
**Pull request by xupyprmv**
Refactoring for #473 `RqHref.java:83-86: RqMethod already validates Request-Line and... #473`

I created RqRequestLine interface that 2 methods:

String requestLineHeader() throws IOException; // returns valid Request-line

String requestLineHeaderToken(Token token) throws IOException; // returns specified token from Request-Line

Also I added class that implements RqRequestLine interface and provides Request-Line validation.

### CODE REVIEW

**Comment by xupyprmv** [comment]
@davvd Please proceed with code review.

**Comment by davvd** [comment]
@xupyprmv Many thanks, I will find someone to review it before we merge

**Comment by davvd** [comment]
@dmzaytsev review this one, please

**Comment by dmzaytsev** [comment]
@xupyprmv pull request description explains the solution proposed and contains **a link** to the original ticket it is related to
the solution should be described more detail

File: src/main/java/org/takes/rq/RqRequestLine.java [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+import org.takes.Request;
+
+/**
+ * HTTP Request-Line parsing.
+ *
+ * <p>All implementations of this interface must be immutable and thread-safe.
+ *
+ * @author Vladimir Maksimenko (xupypr@xupypr.com)
+ * @version $Id$
+ * @since 1.0
```
Comment (dmzaytsev) [inline src/main/java/org/takes/rq/RqRequestLine.java]: @xupyprmv `0.30`

File: src/main/java/org/takes/rq/RqRequestLine.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+    /**
+     * URI token.
+     */
+    int URI = 2;
+
+    /**
+     * HTTPVERSION token.
+     */
+    int HTTPVERSION = 3;
```
Comment (dmzaytsev) [inline src/main/java/org/takes/rq/RqRequestLine.java]: @xupyprmv I think we can use enumeration instead of these constants here

File: src/main/java/org/takes/rq/RqRequestLine.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        @Override
+        public String requestLineHeader() throws IOException {
+            return this.extractFromRequestLineHeader(null);
+        }
+
+        @Override
+        public String requestLineHeaderToken(final int index)
+                throws IOException {
+            // @checkstyle MagicNumberCheck (1 lines)
+            if (index < 1 || index > 3) {
```
Comment (dmzaytsev) [inline src/main/java/org/takes/rq/RqRequestLine.java]: @xupyprmv so if we will use enumeration we don't need this condition

File: src/main/java/org/takes/rq/RqRequestLine.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+         * Ctor.
+         * @param req Original request
+         */
+        public Base(final Request req) {
+            super(req);
+        }
+
+        @Override
+        public String requestLineHeader() throws IOException {
+            return this.extractFromRequestLineHeader(null);
```
Comment (dmzaytsev) [inline src/main/java/org/takes/rq/RqRequestLine.java]: @xupyprmv can we avoid using null here and below?
http://www.yegor256.com/2014/05/13/why-null-is-bad.html

File: src/main/java/org/takes/rq/RqHref.java [hunk 5]
```diff
@@ -80,28 +69,10 @@ public Base(final Request req) {
...
-                REQUEST_PATTERN.matcher(line);
-            if (!matcher.matches()) {
-                throw new HttpException(
-                    HttpURLConnection.HTTP_BAD_REQUEST,
-                    String.format("Illegal Request-Line: %s", line)
-                );
-            }
-            final String uri = matcher.group(2);
+            final String uri = new RqRequestLine.Base(this)
+                    .requestLineHeaderToken(RqRequestLine.URI);
```
Comment (dmzaytsev) [inline src/main/java/org/takes/rq/RqHref.java]: @xupyprmv the indentation step should be 4 spaces

File: src/main/java/org/takes/rq/RqMethod.java [hunk 6]
```diff
@@ -130,14 +121,8 @@ public Base(final Request req) {
...
-            final String line = this.head().iterator().next();
-            final Matcher matcher = PATTERN.matcher(line);
-            if (!matcher.matches()) {
-                throw new IOException(
-                    String.format("Invalid HTTP method line: %s", line)
-                );
-            }
-            final String method = matcher.group(1);
+            final String method = new RqRequestLine.Base(this)
+                    .requestLineHeaderToken(RqRequestLine.METHOD);
```
Comment (dmzaytsev) [inline src/main/java/org/takes/rq/RqMethod.java]: @xupyprmv the indentation step should be 4 spaces

File: src/main/java/org/takes/rq/RqRequestLine.java [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    @EqualsAndHashCode(callSuper = true)
+    final class Base extends RqWrap implements RqRequestLine {
+        /**
+         * HTTP Request-line pattern.
+         * [!-~] is for method or extension-method token (octets 33 - 126).
+         * @see <a href="http://www
+         * .w3.org/Protocols/rfc2616/rfc2616-sec5.html#sec5.1">RFC 2616</a>
+         */
+        private static final Pattern PATTERN = Pattern.compile(
+                "([!-~]+) ([^ ]+)( [^ ]+)?"
```
Comment (dmzaytsev) [inline src/main/java/org/takes/rq/RqRequestLine.java]: @xupyprmv the indentation step should be 4 spaces

**Comment by dmzaytsev** [comment]
@xupyprmv please see a few comments above 
also pay attention the indentation step should be 4 spaces

File: src/main/java/org/takes/rq/RqRequestLine.java (hunk 1 above)
Comment (xupyprmv) [inline src/main/java/org/takes/rq/RqRequestLine.java]: @dmzaytsev I take version 1.0 from `pom.xml` file.  Please, can you explain where I can get information about next version. That helps me not to do the same mistake in future.

Comment (dmzaytsev) [inline src/main/java/org/takes/rq/RqRequestLine.java]: @xupyprmv I add 0.0.1 to the last release usually 
https://github.com/yegor256/takes/releases

Comment (dmzaytsev) [inline src/main/java/org/takes/rq/RqRequestLine.java]: @xupyprmv so it will be `0.29.1`

**Comment by xupyprmv** [comment]
@dmzaytsev Thank you for CR. I refactored code according to your remarks. Please, proceed with CR.

File: src/main/java/org/takes/http/BkReuse.java [hunk 8]
```diff
@@ -1,51 +0,0 @@
...
-import java.io.IOException;
-import java.net.Socket;
-
-/**
- * Reusable back-end.
- *
- * @author Piotr Pradzynski (prondzyn@gmail.com)
- * @version $Id$
- */
-public final class BkReuse extends BkWrap {
```
Comment (dmzaytsev) [inline src/main/java/org/takes/http/BkReuse.java]: @xupyprmv why did you remove this class?

File: src/test/java/org/takes/http/BkReuseTest.java [hunk 9]
```diff
@@ -1,143 +0,0 @@
...
- * @author Piotr Pradzynski (prondzyn@gmail.com)
- * @version $Id$
- * @checkstyle MultipleStringLiteralsCheck (500 lines)
- * @todo #306:30min Implement missing BkReuse.accept method to support
- *  HTTP persistent connections and to pass below three tests. BkReuse.accept
- *  should handles more than one HTTP requests in one connection and return
- *  correct HTTP status when Content-Length is not specified.
- */
-@SuppressWarnings("PMD.AvoidDuplicateLiterals")
-public final class BkReuseTest {
```
Comment (dmzaytsev) [inline src/test/java/org/takes/http/BkReuseTest.java]: @xupyprmv same here
why did you remove it?

**Comment by dmzaytsev** [comment]
@xupyprmv just two questions above

**Comment by xupyprmv** [comment]
@dmzaytsev I rollback removing of `BkReuse` and `BkReuseTest`. It was my fail. It is a part of work for #456, not for current issue.

**Comment by dmzaytsev** [comment]
@xupyprmv no problem
thanks for PR

**Comment by dmzaytsev** [comment]
@rultor merge

File: src/main/java/org/takes/rq/RqRequestLine.java [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+ *
+ * <p>All implementations of this interface must be immutable and thread-safe.
+ *
+ * @author Vladimir Maksimenko (xupypr@xupypr.com)
+ * @version $Id$
+ * @since 0.29.1
+ */
+public interface RqRequestLine extends Request {
+
+    public static enum Token {
```
Comment (yegor256) [inline src/main/java/org/takes/rq/RqRequestLine.java]: @xupyprmv oh, that's a very bad idea... just create methods for each "token" you need to get. Using "tokens" as method configurators is a very bad idea. I will review the rest soon

Comment (xupyprmv) [inline src/main/java/org/takes/rq/RqRequestLine.java]: @yegor256 Do you have any remarks about other parts of PR?

**Comment by xupyprmv** [comment]
@davvd I'm waiting end of @yegor256 CR here.

**Comment by xupyprmv** [comment]
@yegor256 I refactored code according to your remark. Please, proceed with CR. (Sorry, I'm not wait for end of your previous CR, @davvd urges me to close #473 in 2 days)

File: src/main/java/org/takes/rq/RqRequestLine.java [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+     * @author Vladimir Maksimenko (xupypr@xupypr.com)
+     * @version $Id$
+     * @since 1.0
+     */
+    @EqualsAndHashCode(callSuper = true)
+    final class Base extends RqWrap implements RqRequestLine {
+        /**
+         * HTTP Request-line pattern.
+         * [!-~] is for method or extension-method token (octets 33 - 126).
+         * @see <a href="http://www
```
Comment (yegor256) [inline src/main/java/org/takes/rq/RqRequestLine.java]: @xupyprmv this won't work with JavaDoc. just put it all in one long line, checkstyle won't complain

File: src/main/java/org/takes/rq/RqRequestLine.java [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+ * @since 0.29.1
+ */
+public interface RqRequestLine extends Request {
+
+    /**
+     * Get Request-Line header.
+     * @return HTTP Request-Line header
+     * @throws IOException If fails
+     */
+    String requestLineHeader() throws IOException;
```
Comment (yegor256) [inline src/main/java/org/takes/rq/RqRequestLine.java]: @xupyprmv why not just `header`? there are many headers? we really need to say "request line header" in order not to confuse it with some other header? see http://www.yegor256.com/2015/01/12/compound-name-is-code-smell.html keep in mind, the longer the name - the more messy is your code. if you really need long names - it's a sign that your code needs refactoring

File: src/main/java/org/takes/rq/RqRequestLine.java [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+     * @throws IOException If fails
+     */
+    String method() throws IOException;
+
+    /**
+     * Get Request-Line Request-URI token.
+     * @return HTTP Request-Line method token
+     * @throws IOException If fails
+     */
+    String requestUri() throws IOException;
```
Comment (yegor256) [inline src/main/java/org/takes/rq/RqRequestLine.java]: @xupyprmv how about `uri()`?

File: src/main/java/org/takes/rq/RqRequestLine.java [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+
+    /**
+     * Get Request-Line HTTP-Version token.
+     * @return HTTP Request-Line method token
+     * @throws IOException If fails
+     */
+    String httpVersion() throws IOException;
```
Comment (yegor256) [inline src/main/java/org/takes/rq/RqRequestLine.java]: @xupyprmv how about `version()`?

File: src/main/java/org/takes/rq/RqRequestLine.java [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+            return this.requestLineHeaderToken(Token.HTTPVERSION);
+        }
+
+        /**
+         * Get Request-Line header token.
+         * @param token Token
+         * @return HTTP Request-Line header token
+         * @throws IOException If fails
+         */
+        private String requestLineHeaderToken(final Token token)
```
Comment (yegor256) [inline src/main/java/org/takes/rq/RqRequestLine.java]: @xupyprmv what do you think about `token()`?

File: src/main/java/org/takes/rq/RqRequestLine.java [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+            return result.trim();
+        }
+
+        /**
+         * Get Request-Line header.
+         *
+         * @return Valid Request-Line header
+         * @throws IOException If fails
+         */
+        private String getRequestLineHeader() throws IOException {
```
Comment (yegor256) [inline src/main/java/org/takes/rq/RqRequestLine.java]: @xupyprmv maybe just `line()` is good enough?

File: src/main/java/org/takes/rq/RqRequestLine.java [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+        }
+
+        /**
+         * Validate Request-Line according to PATTERN.
+         *
+         * @param requestline Request-Line header
+         * @return Matcher that can be used to extract tokens
+         * @throws HttpException If fails
+         */
+        private Matcher validateRequestLine(final String requestline)
```
Comment (yegor256) [inline src/main/java/org/takes/rq/RqRequestLine.java]: @xupyprmv `requestline` is a compound name, don't use it. just `line` is enough

Comment (yegor256) [inline src/main/java/org/takes/rq/RqRequestLine.java]: @xupyprmv also, read this article: http://www.yegor256.com/2015/12/08/temporal-coupling-between-method-calls.html

Comment (yegor256) [inline src/main/java/org/takes/rq/RqRequestLine.java]: @xupyprmv it's a bad name for the method. it should be called `matcher`, not `validate`

**Comment by yegor256** [comment]
@xupyprmv the code is good, but your naming is rather messy. I posted some comments above. first rule: use single words for method and variable names. second rule: read [this](http://www.yegor256.com/images/books/elegant-objects/seven-pages.pdf)

**Comment by xupyprmv** [comment]
@yegor256 Thanks for advises. I refactored code according to your review. Please proceed with CR.

File: src/main/java/org/takes/rq/RqRequestLine.java [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+         */
+        public Base(final Request req) {
+            super(req);
+        }
+
+        @Override
+        public String header() throws IOException {
+            final String line = this.line();
+            this.matcher(line);
+            return line;
```
Comment (yegor256) [inline src/main/java/org/takes/rq/RqRequestLine.java]: @xupyprmv read this one again: http://www.yegor256.com/2015/12/08/temporal-coupling-between-method-calls.html Your code here should look like this:

```java
public String header() {
  return this.validated(this.line());
}
```

See the idea?

Comment (yegor256) [inline src/main/java/org/takes/rq/RqRequestLine.java]: @xupyprmv maybe I'm wrong with this exact example, but `this.matcher(line)` is a code smell. You should **never** use functions as procedures, ignoring their returns.

**Comment by yegor256** [comment]
@xupyprmv see my comments above, I don't like the design of private methods. they are very "procedural"

**Comment by xupyprmv** [comment]
@yegor256 I refactored code. It is more "functional" now. The reason why I used "procedural" approach before is to avoid "copy-paste style". Because I need `Matcher` for two purposes: to validate Request-Line and to extract groups. But for performance reasons I try to use `match` only once for each call. Please, proceed with CR.

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by yegor256** [comment]
@xupyprmv looks good now, thanks

**Comment by davvd** [comment]
@elenavolokhova please, review this ticket for compliance with our [QA rules](http://at.teamed.io/qa.html)

**Comment by davvd** [comment]
@rultor deploy

**Comment by elenavolokhova** [comment]
@davvd Looks good!

**Comment by davvd** [comment]
> @davvd Looks good!

@elenavolokhova thank you

**Comment by davvd** [comment]
@dmzaytsev just added 10 mins to @elenavolokhova (for QA), payment ID is `74003682`... thanks, I just added **31 mins** to your account, payment `74003698`, 263 hours and 12 mins spent... review comments (c=16) added as a bonus... added +31 to your rating, now it is equal to [+1501](http://www.netbout.com/b/35166?open=rating)
